### PR TITLE
Fix audio tests not checking for unobserved exceptions and not disposing on the audio thread

### PR DESCRIPTION
**Pull request by Susko3**
`TestChannelDoesNotPlayIfReachedEndAndMovedMixers` would fail in `TearDown` locally and on [CI](https://github.com/ppy/osu-framework/actions/runs/5666868487/job/15354483805) when running in release configuration.

Not sure why it would only fail in release and not in debug, but running the update-after-dispose on the audio thread seems more correct, as that's what actually happens in a real [`AudioThread`](https://github.com/ppy/osu-framework/blob/48e9bacb8ec16510bc0c58de5bb22bcfd6f0f336/osu.Framework/Threading/AudioThread.cs#L95-L99).

### CODE REVIEW

**Review by frenzibyte — COMMENTED** [review]
There's much more that can be investigated here which looks serious and may potentially occur in real game scenarios. To start off, running the test in `Release` configuration locally and keeping track of `TrackBass.Mixer` shows that only when running in `Release` configuration, `Mixer` will be set to null before the track updates for the last time.

File: osu.Framework.Tests/Audio/BassTestComponents.cs [hunk 1]
```diff
@@ -93,7 +93,7 @@ public void RunOnAudioThread(Action action)
         public void Dispose()
         {
             allComponents.Dispose();
-            allComponents.Update(); // Actually runs the disposal.
+            Update(); // Actually runs the disposal.
             Bass.Free();
         }
```
[the remark this answers is not in the document]
Comment (Susko3) [reply osu.Framework.Tests/Audio/BassTestComponents.cs:98]: Yeah, that would match [what the audio thread is doing in real usages](https://github.com/ppy/osu-framework/pull/5946#pullrequestreview-1547665885).

**Comment by Susko3** [comment]
I just tried to re-run the test on `master`, and it fails a [`Debug.Assert(CanPerformInline)`](https://github.com/ppy/osu-framework/blob/ba1385330cc501f34937e08257e586c84e35d772/osu.Framework/Audio/Mixing/Bass/BassAudioMixer.cs#L71), but the test says 'success'.
You can see the assert failing with Rider's 'user-unhandled exception'.

Minimum repro case for this behaviour. This test will **pass** on my machine:

```cs
[TestFixture]
public class TaskAssertTest
{
    [Test]
    public void Test()
    {
        var t = new Task(() => Debug.Assert(false));
        t.RunSynchronously();
    }
}
```

Also, any exceptions that happen inside `RunOnAudioThread` also won't propagate outwards.

**Comment by Susko3** [comment]
The following diff makes the test "pass" in release on `master`:

```diff
diff --git a/osu.Framework/Audio/Mixing/Bass/BassAudioMixer.cs b/osu.Framework/Audio/Mixing/Bass/BassAudioMixer.cs
index 37ed04d6d..d4fc243b8 100644
--- a/osu.Framework/Audio/Mixing/Bass/BassAudioMixer.cs
+++ b/osu.Framework/Audio/Mixing/Bass/BassAudioMixer.cs
@@ -68,7 +68,7 @@ protected override void AddInternal(IAudioChannel channel)
 
         protected override void RemoveInternal(IAudioChannel channel)
         {
-            Debug.Assert(CanPerformInline);
+            if (!CanPerformInline) throw new InvalidOperationException();
 
             if (!(channel is IBassAudioChannel bassChannel))
                 return;
```

> keeping track of TrackBass.Mixer shows that only when running in Release configuration, Mixer will be set to null before the track updates for the last time.

In Debug, the assert aborts execution at `RemoveInternal`, [just before `channel.Mixer` is set to `null`](https://github.com/ppy/osu-framework/blob/ba1385330cc501f34937e08257e586c84e35d772/osu.Framework/Audio/Mixing/AudioMixer.cs#L68). The same is not true for Release, as the assert gets compiled-out.

## Conclusion

`osu.Framework.Tests.Audio` are invalid because they don't check for unobserved exceptions in actions queued by [`EnqueueAction`](https://github.com/ppy/osu-framework/blob/ba1385330cc501f34937e08257e586c84e35d772/osu.Framework/Audio/AudioComponent.cs#L38) and the [`RunOnAudioThread`](https://github.com/ppy/osu-framework/blob/75ed421f60228920abd819cebc5982cb7799bbbb/osu.Framework.Tests/Audio/BassTestComponents.cs#L72) test helper method. This hides an actual problem with the tests, namely that [`BassTestComponents`](https://github.com/ppy/osu-framework/blob/75ed421f60228920abd819cebc5982cb7799bbbb/osu.Framework.Tests/Audio/BassTestComponents.cs#L93-L98) doesn't properly execute the disposal on the audio thread.

The problem is only visible with `TestChannelDoesNotPlayIfReachedEndAndMovedMixers`. Only this test throws an unobserved exception (from `Debug.Assert`, so it would throw only on debug builds). I've not investigated why only this test throws.

**Review by smoogipoo — APPROVED** [review]
Looks okay to me

**Pull request merged**